src/swap/defi/rango.ts

```diff
@@ -0,0 +1,542 @@
+    const toAddress = await getAddress(toWallet)
+
+    const fromMainnetCode =
+      MAINNET_CODE_TRANSCRIPTION[fromWallet.currencyInfo.pluginId]
+    const toMainnetCode =
+      MAINNET_CODE_TRANSCRIPTION[toWallet.currencyInfo.pluginId]
+
+    if (fromMainnetCode === '' || toMainnetCode === '') {
+      throw new SwapCurrencyError(swapInfo, request)
+    }
+
+    const now = Date.now()
+    if (
+      now - exchangeInfoLastUpdate > EXCHANGE_INFO_UPDATE_FREQ_MS ||
+      exchangeInfo == null
+    ) {
+      try {
+        const exchangeInfoResponse = await promiseWithTimeout(
+          fetchInfo(fetchCors, `v1/exchangeInfo/${appId}`)
+        )
+
+        if (exchangeInfoResponse.ok === true) {
+          exchangeInfo = asExchangeInfo(await exchangeInfoResponse.json())
+          exchangeInfoLastUpdate = now
+        } else {
+          // Error is ok. We just use defaults
+          const text: string = await exchangeInfoResponse.text()
+          log.warn(
+            `Error getting info server exchangeInfo. Using defaults... Error: ${text}`
+          )
+        }
+      } catch (e: any) {
+        log.warn(
+          'Error getting info server exchangeInfo. Using defaults...',
+          e.message
+        )
+      }
+    }
+
+    const fromSymbol =
+      rangoTokens?.[fromMainnetCode]?.[fromContractAddress.toLowerCase()]
+    const toSymbol =
+      rangoTokens?.[toMainnetCode]?.[toContractAddress.toLowerCase()]
+
+    if (fromSymbol === undefined || toSymbol === undefined) {
+      throw new Error('Rango could not proceed with the exchange')
+    }
+
+    if (exchangeInfo != null) {
+      const { rango } = exchangeInfo.swap.plugins
+      rangoServers = rango?.rangoServers ?? rangoServers
+    }
+
+    const swapParameters = {
+      apiKey: rangoApiKey,
+      from: assetToString({
+        blockchain: fromMainnetCode,
+        address: fromContractAddress,
+        symbol: fromSymbol
+      }),
+      to: assetToString({
+        blockchain: toMainnetCode,
+        address: toContractAddress,
+        symbol: toSymbol
+      }),
+      fromAddress: fromAddress,
+      toAddress: toAddress,
+      amount: nativeAmount,
+      disableEstimate: true,
+      slippage: DEFAULT_SLIPPAGE,
+      ...(referrerAddress !== undefined &&
+        referrerAddress !== '' &&
+        referrerFee !== undefined &&
+        referrerFee !== '' && { referrerAddress, referrerFee })
+    }
+
+    const params = makeQueryParams(swapParameters)
+
+    const swapResponse = await fetchWaterfall(
+      fetchCors,
+      rangoServers,
+      `basic/swap?${params}`,
+      {
+        headers
+      }
+    )
+
+    if (!swapResponse.ok) {
+      const responseText = await swapResponse.text()
+      throw new Error(`Rango could not fetch quote: ${responseText}`)
+    }
+
+    const swap = asSwapResponse(await swapResponse.json())
+    const { route, tx } = swap
+
+    if (swap.resultType !== 'OK') {
+      throw new Error(
```

Comment:
Is there any better info that could be determined? We have specific errors for 

1. SwapBelowLimitError
2. SwapCurrencyError (currency unsupported)
3. SwapPermissionError (region, verification, etc)
4. SwapAboveLimitError